internal/scheduler/scheduler.go

```diff
@@ -0,0 +1,181 @@
+
+// GetItemsFunc is a callback that returns items for the scheduler to consider.
+type GetItemsFunc func(ctx context.Context, now time.Time) []Item
+
+// Item is a describes item that can be scheduled with a function invoked the next time the item is due.
+type Item struct {
+	Description string
+	NextTime    time.Time
+	Trigger     func()
+}
+
+// Scheduler is a scheduler triggers the next upcoming events based on the list of items returned by listItems().
+type Scheduler struct {
+	TimeNow func() time.Time
+	Debug   bool
+
+	refreshRequested chan string
+	getItems         GetItemsFunc
+	closed           chan struct{}
+	wg               sync.WaitGroup
+}
+
+// Options provides options for the scheduler.
+type Options struct {
+	TimeNow        func() time.Time
+	Debug          bool
+	RefreshChannel chan string
+}
+
+// Start runs a new scheduler that will call getItems() to get the list of items to schedule.
+func Start(ctx context.Context, getItems GetItemsFunc, opts Options) *Scheduler {
+	timeNow := opts.TimeNow
+
+	if timeNow == nil {
+		timeNow = clock.Now
+	}
+
+	s := &Scheduler{
+		TimeNow:          timeNow,
+		refreshRequested: opts.RefreshChannel,
+		closed:           make(chan struct{}),
+		getItems:         getItems,
+		Debug:            opts.Debug,
+	}
+
+	s.wg.Add(1)
+
+	go func() {
+		defer s.wg.Done()
+
+		s.run(ctxutil.Detach(ctx))
+	}()
+
+	return s
+}
+
+const sleepTimeWhenNoUpcomingSnapshots = 8 * time.Hour
+
+func (s *Scheduler) upcomingItems(ctx context.Context, now time.Time) (nextTriggerTime time.Time, toTrigger []Item) {
+	allsm := s.getItems(ctx, now)
+
+	for _, t := range allsm {
+		if nextTriggerTime.IsZero() || t.NextTime.Before(nextTriggerTime) {
+			nextTriggerTime = t.NextTime
+			toTrigger = nil
+		}
+
+		if t.NextTime.Equal(nextTriggerTime) {
+			toTrigger = append(toTrigger, t)
+		}
+	}
+
+	if s.Debug {
+		log(ctx).Debugf(">>>> scheduling %v items (now %v)", len(allsm), now.Format(time.RFC3339))
+
+		for _, it := range allsm {
+			log(ctx).Debugf("  %v %v", it.NextTime.Format(time.RFC3339), it.Description)
+		}
+	}
+
+	return nextTriggerTime, toTrigger
+}
+
+func sleepTimeOrDefault(now, t time.Time, def time.Duration) time.Duration {
+	if t.IsZero() {
+		return def
+	}
+
+	return t.Sub(now)
+}
+
+// Stop stops the scheduler.
+func (s *Scheduler) Stop() {
+	close(s.closed)
+	s.wg.Wait()
+}
+
+func (s *Scheduler) run(ctx context.Context) {
+	for {
+		now := s.TimeNow()
+		nextTriggerTime, toTrigger := s.upcomingItems(ctx, now)
+
+		sleepTimeUntilNextTrigger := sleepTimeOrDefault(now, nextTriggerTime, sleepTimeWhenNoUpcomingSnapshots)
+		if sleepTimeUntilNextTrigger < 0 {
+			sleepTimeUntilNextTrigger = 0
+		}
+
+		if s.Debug {
+			log(ctx).Debugf("sleeping for %v until %v (%v)",
+				sleepTimeUntilNextTrigger,
+				nextTriggerTime.Format(time.RFC3339),
+				TriggerNames(toTrigger),
+			)
+		}
+
+		select {
+		case <-s.closed:
+			// stopping, just exit
+			return
+
+		case <-time.After(sleepTimeUntilNextTrigger):
```

Comment:
This is a potential memory-leak if we end up triggering too many refreshes. Given the potential wait times could be hours, this could accumulate easily:

```
The underlying Timer is not recovered by the garbage collector
 until the timer fires. If efficiency is a concern, use `NewTimer`
 instead and call `Timer.Stop` if the timer is no longer needed.
```
REF: https://pkg.go.dev/time#After

You can use Timer.Reset() to set a new wait.

Reply:
done in base PR.